### PR TITLE
[WIP] Add EDM Pipeline and Scheduler Implementation

### PR DESCRIPTION
**Pull request by dg845**
# What does this PR do?

This PR adds a new pipeline and scheduler to implement the EDM model of Karras et at. 2022 ([paper](https://arxiv.org/pdf/2206.00364.pdf), [code](https://github.com/NVlabs/edm)).

edm_paper_table_1

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@patrickvonplaten
@williamberman
@ayushtues

### CODE REVIEW

**Comment by dg845** [comment]
@patrickvonplaten and/or @williamberman, would you be able to do a preliminary review of the design of `KarrasEDMPipeline` and `KarrasEDMScheduler`? The goal of this PR is to have a pipeline/scheduler that exactly replicates the EDM model; I believe other pipelines/schedulers do not reproduce the EDM model because they do not use the same preconditioning (e.g. `KarrasVePipeline`/`KarrasVeScheduler`, `HeunDiscreteScheduler`). Design notes are below.

----

When I was working on `KarrasEDMPipeline`, I realized that the EDM model is essentially almost entirely implemented by [`KarrasVePipeline`](https://github.com/huggingface/diffusers/blob/v0.19.3/src/diffusers/pipelines/stochastic_karras_ve/pipeline_stochastic_karras_ve.py#L25) + [`KarrasVeScheduler`](https://github.com/huggingface/diffusers/blob/v0.19.3/src/diffusers/schedulers/scheduling_karras_ve.py#L48). The only difference as far as I know is that `KarrasVePipeline` uses the preconditioning scalings corresponding to the "VE" model (see Equation 213 of the [paper](https://arxiv.org/pdf/2206.00364.pdf)) with all other design choices matching that of the "EDM" column of Table 1 (reproduced in https://github.com/huggingface/diffusers/pull/4481#issue-1837610844). Thus the design of `KarrasVePipeline`/`KarrasEDMPipeline` and `KarrasVeScheduler`/`KarrasEDMScheduler` are very similar, with differences as follows:

Differences between `KarrasVePipeline`/`KarrasEDMPipeline`
- As discussed above, the preconditioning scalings are changed to match that of the EDM model in `KarrasEDMPipeline`. The scalings are not hardcoded but call methods of `KarrasEDMScheduler`.
- Providing class labels to `KarrasEDMPipeline.__call__` is supported for class-conditional models.
- Some refactoring changes have been made with methods like `prepare_latents`, `post_process_image`, `check_inputs`, etc.

Differences between `KarrasVeScheduler`/`KarrasEDMScheduler`

- `KarrasEDMScheduler` has methods like precondition_noise and precondition_inputs to explicitly handle preconditioning as described in Section 5 of the [paper](https://arxiv.org/pdf/2206.00364.pdf). The implementation of these methods allow them to be potentially be extended to other samplers in the Karras framework, like the Karras VE and VP model, although only the EDM preconditioning is implemented for now.

One possible design is to have one pipeline + scheduler that has a sampling loop following `KarrasVePipeline`/`KarrasEDMPipeline`:

1. Temporarily raise `sample` to noise level `sigma_hat` to get a new noisy sample `sample_hat`. [`scheduler.add_noise_to_input`]
2. Precondition the U-Net inputs.
3. Evaluate the U-Net at noisy sample `sample_hat` and noise level `sigma_hat`.
4. Get the denoising prediction and take a Euler step at `sigma_hat`. [`scheduler.step`]
5. If applying a 2nd order correction evaluate the U-Net again at the sample produced from the Euler step and noise level `sigma_prev` and take a Heun step at `sigma_prev`. [`scheduler.step_correct`].

This is what is currently implemented in the PR.

IIRC the design for `KarrasVePipeline`/`KarrasVeScheduler` is considered deprecated, so perhaps the above design might not be the best. I think it might also be possible to support the EDM model through modifying [`HeunDiscreteScheduler`](https://github.com/huggingface/diffusers/blob/v0.19.3/src/diffusers/schedulers/scheduling_heun_discrete.py#L71). In this case I think we would still need a new `EDMPipeline` to add noise to the sample and perform input preconditioning, which aren't currently fully supported by `HeunDiscreteScheduler`, and then modify `HeunDiscreteScheduler` to perform output preconditioning, perhaps by adding a new `"sample_precondition"` `prediction_type`:

```python
    def step(...):
        ...
        elif self.config.prediction_type == "sample_precondition":
            c_skip = sigma_data**2 / (sigma**2 + sigma_data**2)
            c_out = sigma * sigma_data / (sigma**2 + sigma_data**2) ** 0.5
            pred_original_sample = c_skip * sample + c_out * model_output
        ...
```
I think this should work whether `HeunDiscreteScheduler` is in first order or second order mode (e.g. processing the model_output for the Euler step or the 2nd order correction step). (This is not implemented currently.)

**Comment by dg845** [comment]
(As an aside, supporting temporarily increasing the noise level for EDM stochastic sampling in `HeunDiscreteScheduler` would require a `timestep` equivalent of `scheduler.scale_model_input`:

https://github.com/huggingface/diffusers/blob/de9c72d58c3eb4bcd50901d0c23c7c46f0cec29b/src/diffusers/schedulers/scheduling_heun_discrete.py#L341-L345

I assume changing the API to add this method would probably necessitate changes to a bunch of pipelines.)

File: src/diffusers/schedulers/scheduling_edm.py [hunk 1]
```diff
@@ -0,0 +1,488 @@
...
+        max_inv_rho = sigma_max ** (1 / rho)
+        sigmas = (max_inv_rho + ramp * (min_inv_rho - max_inv_rho)) ** rho
+        return sigmas
+
+    def step(
+        self,
+        model_output: torch.FloatTensor,
+        sigma_hat: float,
+        sample_hat: torch.FloatTensor,
+        sigma_prev: float,
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_edm.py:323]:
```suggestion
```
We should not take any sigmas as input s here - this breaks our design with other schedulers. Can you try to change the design to adhere to: https://github.com/huggingface/diffusers/blob/d185b5ed5f23c5912918ee81881a3c03f9359523/src/diffusers/schedulers/scheduling_lms_discrete.py#L318 so that this scheduler can also be used with other stable diffusion pipelines? 

Note that this scheduler: https://github.com/huggingface/diffusers/blob/main/src/diffusers/schedulers/scheduling_karras_ve.py is outdated/deprecated. We should probably either delete it or deprecate it (cc @yiyixuxu )

**Review by patrickvonplaten — COMMENTED** [review]
Looks very nice to me now! Thanks for making the scheduler refactor! Could we make add some fast tests for the EDM scheduler?

**Comment by dg845** [comment]
> Could we make add some fast tests for the EDM scheduler?

Currently working on it :).

**Comment by dg845** [comment]
Some notes about the current design: currently the new scheduler `KarrasEDMScheduler` is very similar to [`HeunDiscreteScheduler`](https://github.com/huggingface/diffusers/blob/main/src/diffusers/schedulers/scheduling_heun_discrete.py), in that they are both `order = 2` schedulers that do Euler and Heun updates on alternating calls to `step(...)`. The main differences between the two are as follows:

- `KarrasEDMScheduler` supports `gamma > 0` (stochastic sampling) via returning both a scaled model input and scaled timestep from `scale_model_input`. This diverges from the normal scheduler API and breaks compatibility with most pipelines (since the return type of `scale_model_input` is now a tuple). [Perhaps one way around this is to calculate the `sigma_hat` schedule in advance - I think this could be possible because each `sigma_hat` only depends on its corresponding `sigma` and `num_inference_steps`, `s_tmin`, `s_tmax`, and `s_churn`, all of which are known at the time we call `set_timesteps(...)`.]
- `KarrasEDMPipeline` supports noise preconditioning of the timestep and output preconditioning (skip/output scaling), whereas `HeunDiscreteScheduler` does not. (See Table 1 above; `HeunDiscreteScheduler` does do input preconditioning inside its `scale_model_input` method).
- `KarrasEDMPipeline` supports custom `timesteps` in `set_timesteps`. I think this is important because the [EDM paper](https://arxiv.org/pdf/2206.00364.pdf) and papers building off it (such as the [consistency models paper](https://arxiv.org/pdf/2303.01469.pdf)), tend to work purely with `sigma` schedules. I'm not aware if there is a good way to convert between `beta` schedules and `sigma` schedules (intuitively, my guess is that a `beta` schedule does not uniquely determine a `sigma` schedule and vice versa), so being able to supply custom `timesteps` makes it much easier to reproduce the `sigma` schedules used in the papers.

An alternative would be to fold the above changes directly into `HeunDiscreteScheduler` (especially if it's possible to support `gamma` values greater than `0` within the current scheduler API).

File: src/diffusers/schedulers/scheduling_edm.py [hunk 2]
```diff
@@ -0,0 +1,711 @@
...
+
+    betas = []
+    for i in range(num_diffusion_timesteps):
+        t1 = i / num_diffusion_timesteps
+        t2 = (i + 1) / num_diffusion_timesteps
+        betas.append(min(1 - alpha_bar_fn(t2) / alpha_bar_fn(t1), max_beta))
+    return torch.tensor(betas, dtype=torch.float32)
+
+
+class KarrasEDMScheduler(SchedulerMixin, ConfigMixin):
```
Comment (williamberman) [inline src/diffusers/schedulers/scheduling_edm.py:95]: Doesn't the karras paper just use existing ode methods? I.e. is uses heun and euler iirc? And I think we already support the karras sigmas on existing schedulers. Would we be potentially able to use existing schedulers?

Comment (williamberman) [reply src/diffusers/schedulers/scheduling_edm.py:95]: There's also the existing DPMSolverSDEScheduler which says it implements the sde sampler from the karras paper

Comment (dg845) [reply src/diffusers/schedulers/scheduling_edm.py:95]: > Would we be potentially able to use existing schedulers?

I've created a draft PR that modifies `HeunDiscreteScheduler` instead of creating a new scheduler: #4930. It's pretty similar to `KarrasEDMScheduler` in this PR, so if the modified `HeunDiscreteScheduler` design there looks good we could use that instead of adding a new scheduler.

Regarding `DPMSolverSDEScheduler`, I'm not familiar enough with SDEs to say whether it could be modified, but if it's equivalent to [`sample_dpm_2`](https://github.com/crowsonkb/k-diffusion/blob/master/k_diffusion/sampling.py#L188) in [k-diffusion](https://github.com/crowsonkb/k-diffusion/tree/master) it seems like it's not equivalent to the Karras Heun algorithm since the 2nd order step is different.

**Comment by dg845** [comment]
I have added fast tests for `KarrasEDMScheduler`, which should be passing (they pass locally for me).

File: src/diffusers/pipelines/edm/pipeline_edm.py [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+        hook = None
+        for cpu_offloaded_model in [self.unet]:
+            _, hook = cpu_offload_with_hook(cpu_offloaded_model, device, prev_module_hook=hook)
+
+        if self.safety_checker is not None:
+            _, hook = cpu_offload_with_hook(self.safety_checker, device, prev_module_hook=hook)
+
+        # We'll offload the last model manually.
+        self.final_offload_hook = hook
+
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/edm/pipeline_edm.py:107]:
```suggestion
```

Comment (patrickvonplaten) [reply src/diffusers/pipelines/edm/pipeline_edm.py:107]: This is part of the "base" diffusion pipeline class now

**Comment by patrickvonplaten** [comment]
@dg845 does the scheduler work with Stable Diffusion models as well?

**Comment by dg845** [comment]
I haven't tested it with Stable Diffusion pipelines yet, but the API should be compatible with two caveats:

1. Because stochastic sampling from the temporarily raised noise level happens in `KarrasEDMScheduler.scale_model_input`, it takes an extra optional `generator` argument for reproducibility. This should still work with the `self.scheduler.scale_model_input` call in e.g. `StableDiffusionPipeline`:

https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py#L677

but the results won't be reproducible with the current `StableDiffusionPipeline.__call__` implementation.

2. Only `sigma` schedules are supported currently, so it will be hard to use common Stable Diffusion `beta` schedules for inference.

For (2), beta schedule support shouldn't be too difficult (I've done most of the work to support them over in the other PR that modifies `HeunDiscreteScheduler`) but I'm not sure if there's a good way to get around (1) with only modifications to the scheduler code.

**Comment by patrickvonplaten** [comment]
@yiyixuxu can you take a look here?

**Comment by yiyixuxu** [comment]
gentle pin @dg845 here:) - we would like to have this scheduler very much! let's:
1. remove the pipeline, I think we only need the scheduler
2. test the scheduler with SD pipelines and see some results :)

thanks!
YiYi

**Comment by dg845** [comment]
I have added some fast tests testing the integration between `StableDiffusionPipeline` and `KarrasEDMScheduler`, which pass. However, I've only been able to generate noise from [`runwayml/stable-diffusion-v1-5`](https://huggingface.co/runwayml/stable-diffusion-v1-5) with `KarrasEDMScheduler` so far:

```python
import torch
from diffusers import StableDiffusionPipeline, KarrasEDMScheduler

model_id = "runwayml/stable-diffusion-v1-5"
pipe = StableDiffusionPipeline.from_pretrained(model_id, torch_dtype=torch.float16)
num_train_timesteps = pipe.scheduler.config.num_train_timesteps
prediction_type = pipe.scheduler.config.prediction_type
edm_scheduler = KarrasEDMScheduler(
    num_train_timesteps=num_train_timesteps,
    prediction_type=prediction_type,
    precondition_type=None,
    sigma_min=0.05,
    sigma_max=999.0,
    s_churn=0.0,
    s_noise=1.0,
)
pipe.scheduler = edm_scheduler
pipe = pipe.to(dtype=torch.float16, device="cuda")

prompt = "a photo of an astronaut riding a horse on mars"
generator = torch.manual_seed(349857)
image = pipe(prompt, num_inference_steps=40, generator=generator).images[0]

image.save("astronaut_rides_horse.png")
```

astronaut_rides_horse

To make sure the scheduler is working as expected, I am testing it with EDM model checkpoints and `KarrasEDMPipeline` (specifically, with the [consistency models repo](https://github.com/openai/consistency_models) EDM checkpoints and implementation). The fast pipeline tests in `/tests/pipelines/edm/test_edm.py` are passing, but the slow tests aren't passing yet. A sample image generated from the [edm_imagenet64_ema](https://huggingface.co/dg845/diffusers-cm_edm_imagenet64_ema) checkpoint with `KarrasEDMScheduler` is as follows:

```python
import torch
from diffusers import KarrasEDMPipeline

device = "cuda"
model_id_or_path = "dg845/diffusers-cm_edm_imagenet64_ema"
pipe = KarrasEDMPipeline.from_pretrained(model_id_or_path)
pipe.to(device)

generator = torch.manual_seed(349857)
# ImageNet-64 class label 145 corresponds to king penguins
image = pipe(num_inference_steps=40, class_labels=145, generator=generator).images[0]

image.save("imagenet64_sample.png")
```

imagenet64_sample

I will do some further debugging to see if there are still some bugs in the scheduler.

**Comment by YichengDWu** [comment]
I would like to quote the official recommendation from NVIDIA on how to design a Pipeline that includes preconditioning:

> The clean way to achieve this is to keep denoise unchanged from the viewpoint of external callers (the ODE solver and the training loop), but change the way it utilizes the network internally. Isolate the actual raw network layers into their own black box module net, and wrap it with magnitude management code (“preconditioning”) within denoise.

[source link](https://developer.nvidia.com/blog/generative-ai-research-spotlight-demystifying-diffusion-based-models/)

**Comment by yiyixuxu** [comment]
@dg845 
I will reopen this one if you are still working on this, let us know!
